src/boundary.rs

```diff
@@ -0,0 +1,743 @@
+//! Algorithms for handling port boundaries in a graph.
+
+use std::borrow::Cow;
+use std::collections::{BTreeMap, BTreeSet, HashSet};
+
+use itertools::Itertools;
+
+use crate::algorithms::TopoSort;
+use crate::{Direction, LinkView, NodeIndex, PortIndex, PortView};
+
+/// A port boundary in a graph.
+///
+/// Defined from a set of incoming and outgoing ports.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Boundary {
+    /// The ordered list of incoming ports in the boundary.
+    inputs: Vec<PortIndex>,
+    /// The ordered list of outgoing ports in the boundary.
+    outputs: Vec<PortIndex>,
+}
+
+/// Boundary port ID.
+///
+/// See [`Boundary`] for more information.
+///
+/// The corresponding [`PortIndex`] in the boundary can be retrieved with
+/// [`Boundary::port_index`].
+#[derive(Debug, Clone, Default, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct BoundaryPort {
+    /// The index of the port in the boundary.
+    index: usize,
+    /// The direction of the port.
+    direction: Direction,
+}
+
+/// Trait for graph structures that define a boundary of input and output ports.
+pub trait HasBoundary {
+    /// Returns the boundary of the node.
+    fn port_boundary(&self) -> Cow<'_, Boundary>;
+}
+
+impl Boundary {
+    /// Creates a new boundary from the given input and output ports.
+    ///
+    /// Queries the direction of each port to split them into inputs and outputs.
+    /// For a version that doesn't borrow the graph, use [`Boundary::new`].
+    pub fn from_ports(graph: &impl PortView, ports: impl IntoIterator<Item = PortIndex>) -> Self {
+        let (inputs, outputs): (Vec<_>, Vec<_>) = ports.into_iter().partition_map(|p| match graph
+            .port_direction(p)
+        {
+            Some(Direction::Incoming) => itertools::Either::Left(p),
+            _ => itertools::Either::Right(p),
+        });
```

Comment:
An invalid port should panic, rather than fail silently:
```suggestion
        let (inputs, outputs): (Vec<_>, Vec<_>) = ports.into_iter().partition_map(|p| match graph
            .port_direction(p).unwrap()
        {
            Direction::Incoming => itertools::Either::Left(p),
            _ => itertools::Either::Right(p),
        });
```

Reply:
You might consider adding a `Direction::to_either(&self) -> itertools::Either` function for `Direction`. It seems this patterns comes up a lot